Error message ascii art pointer / arrow points to wrong argument

Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8b21d823468da1cb8ba7442d8b4e0e85

```rust
fn foo(x: u32, a: impl Into<String>, y: u32, b: impl Into<String>) {
    println!("fox: a={}, b={}", a.into(), b.into());
}

fn main() {
    let bar: String = "bar".to_string();
    let baz: &str = "baz";

    for (a, b) in &[(&bar, baz)] {
        let a: &String = a;
        foo(42, a, 43, b);
    }
}
```

The current output is:

```
/Users/mkm/.cargo/bin/cargo play --cargo-option=--color=always /var/folders/zp/pk0v2q4x39z43k91vst1_jsc0000gn/T/scratch.rs
   Compiling p3ofvf9avhdc8sbzh9blgjf7wwnoz v0.1.0 (/var/folders/zp/pk0v2q4x39z43k91vst1_jsc0000gn/T/cargo-play.3oFVf9AvHdC8sbzH9BLGJf7wwnoz)
error[E0277]: the trait bound `String: From<&&str>` is not satisfied
  --> src/main.rs:11:17
   |
1  | fn foo(x: u32, a: impl Into<String>, y: u32, b: impl Into<String>) {
   |                                                      ------------ required by this bound in `foo`
...
11 |         foo(42, a, 43, b);
   |                 ^ the trait `From<&&str>` is not implemented for `String`
   |
   = help: the following implementations were found:
             <String as From<&String>>
             <String as From<&mut str>>
             <String as From<&str>>
             <String as From<Box<str>>>
           and 2 others
   = note: required because of the requirements on the impl of `Into<String>` for `&&str`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `p3ofvf9avhdc8sbzh9blgjf7wwnoz`

To learn more, run the command again with --verbose.

```

Ideally the output should look like:

```
/Users/mkm/.cargo/bin/cargo play --cargo-option=--color=always /var/folders/zp/pk0v2q4x39z43k91vst1_jsc0000gn/T/scratch.rs
   Compiling p3ofvf9avhdc8sbzh9blgjf7wwnoz v0.1.0 (/var/folders/zp/pk0v2q4x39z43k91vst1_jsc0000gn/T/cargo-play.3oFVf9AvHdC8sbzH9BLGJf7wwnoz)
error[E0277]: the trait bound `String: From<&&str>` is not satisfied
  --> src/main.rs:11:17
   |
1  | fn foo(x: u32, a: impl Into<String>, y: u32, b: impl Into<String>) {
   |                                                      ------------ required by this bound in `foo`
...
11 |         foo(42, a, 43, b);
   |                        ^ the trait `From<&&str>` is not implemented for `String`
   |
   = help: the following implementations were found:
             <String as From<&String>>
             <String as From<&mut str>>
             <String as From<&str>>
             <String as From<Box<str>>>
           and 2 others
   = note: required because of the requirements on the impl of `Into<String>` for `&&str`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `p3ofvf9avhdc8sbzh9blgjf7wwnoz`

To learn more, run the command again with --verbose.
```

I tried on stable (1.52.1) , beta and nightly (on rust playground and locally).

This bug seems like a weird mix of things... the bug in question is present here:

https://github.com/rust-lang/rust/blob/9814e830942dcc65e69a0d077cb2e013b5948795/compiler/rustc_typeck/src/check/fn_ctxt/checks.rs#L911-L925

It looks like `ty.walk().any(|arg| arg == predicate.self_ty().into())` is searching for the predicate `<std::string::String as std::convert::From<&&str>>`, but the way it's searching for it is by looking for references to `std::string::String` (the `self_ty` of the predicate).

But since the argument types look like:
```
(u32, &std::string::String, u32, &&str)
```

it means that walking each type and looking for `std::string::String` will only match against it for `&std::string::String`... but the problem is, I don't think the `self_ty` of the predicate is the item of interest? This implementation seems like a heuristic at best and I'm not sure I'm familiar enough with trait resolution internals to handle this problem.

But the steps seem to be something like:

1. Compiler sees that the argument must implement `Into<String>`. It goes "ok for T to implement Into<String>, String must implement From<T>"
2. So the compiler sees `&&str` being passed as `Into<String>`, and so it goes "ok does 'String' implement From<&&str>" and it goes: no, no it doesn't
3. And it goes "ok fire off a diagnostic for a `String` argument not implementing `From<&&str>`"
4. When it tries to fire off that diagnostic, it sees that there is, in fact, a `String` in one of the arguments, it goes, ah! this must be the `String` that doesn't implement `From<&&str>` and fires off the lint pointing at the wrong argument.

My best guess on how to fix this, which I will look into more tomorrow, is to modify step 3 to look more like "ok so `String` doesn't implement `From<&&str>`, why did I care about that again?" and it goes "ah yes in order for `&&str` to satisfy `Into<String>`, fire off a diagnostic for an `&&str` that it doesn't implement `Into<String>`".

I think part of the problem with that implementation is that like....

```
^ the trait `From<&&str>` is not implemented for `String`
```

this text would likely be modified by that? And I'm not certain if that's preferable. I personally don't find it all that great for it to resolve through.

This is the full diagnostic, for reference:

```
error[E0277]: the trait bound `String: From<&&str>` is not satisfied
  --> src/main.rs:21:17
   |
1  | fn foo(x: u32, a: impl Into<String>, y: u32, b: impl Into<String>) {
   |                                                      ------------ required by this bound in `foo`
...
21 |         foo(42, a, 43, b);
   |                 ^ the trait `From<&&str>` is not implemented for `String`
   |
   = help: the following implementations were found:
             <String as From<&String>>
             <String as From<&mut str>>
             <String as From<&str>>
             <String as From<Box<str>>>
           and 2 others
   = note: required because of the requirements on the impl of `Into<String>` for `&&str`
```

While I *do* think this is good for expressing requirement chains, I don't feel the primary label is all that user-friendly. I've seen a lot of beginners go "doesn't implement `From<&&str>`" but it doesn't need that, this is supposed to be `Into<String>`"? Especially when it comes to 3rd party crates, where such a trait-chain would be less familiar/obvious.

--- a/tests/ui/traits/bad-error-message-impl-trait-double-ref-issue-85681.rs
+++ b/tests/ui/traits/bad-error-message-impl-trait-double-ref-issue-85681.rs
@@ -1,0 +1,13 @@
+fn foo(_x: u32, a: impl Into<String>, _y: u32, b: impl Into<String>) {
+    println!("fox: a={}, b={}", a.into(), b.into());
+}
+
+fn main() {
+    let bar: String = "bar".to_string();
+    let baz: &str = "baz";
+
+    for (a, b) in &[(&bar, baz)] {
+        let a: &String = a;
+        foo(42, a, 43, b); //~ ERROR: the trait bound `String: From<&&str>` is not satisfied [E0277]
+    }
+}

--- a/tests/ui/traits/bad-error-message-impl-trait-double-ref-issue-85681.stderr
+++ b/tests/ui/traits/bad-error-message-impl-trait-double-ref-issue-85681.stderr
@@ -1,0 +1,23 @@
+error[E0277]: the trait bound `String: From<&&str>` is not satisfied
+  --> $DIR/bad-error-message-impl-trait-double-ref-issue-85681.rs:11:24
+   |
+LL |         foo(42, a, 43, b);
+   |         ---            ^ the trait `From<&&str>` is not implemented for `String`
+   |         |
+   |         required by a bound introduced by this call
+   |
+   = help: consider casting the `&&str` value to `&str`
+   = note: required for `&&str` to implement `Into<String>`
+note: required by a bound in `foo`
+  --> $DIR/bad-error-message-impl-trait-double-ref-issue-85681.rs:1:56
+   |
+LL | fn foo(_x: u32, a: impl Into<String>, _y: u32, b: impl Into<String>) {
+   |                                                        ^^^^^^^^^^^^ required by this bound in `foo`
+help: consider dereferencing here
+   |
+LL |         foo(42, a, 43, *b);
+   |                        +
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

